### DISCUSSION TITLE
Handling of negative  `mass2`/`tau2` and boosts: Add warnings or exceptions?

### DISCUSSION THREAD

**Discussion by meliache** [scikit-hep/vector#171] [Ideas]
I reconstruct a virtual W four-vector from the difference of reconstructed Bsig and D* four-vectors and was surprised that sometimes, the result of the `mass2` method is negative.
``` python
p = vector.obj(px=-1.1764122315216456, py=0.7286259979009628, pz=1.8719534137922902, E=2.3187584307488134)
print(f"{p.mass=}\n{p.mass2=}\n{p.mass**2=}")
```
which prints
```
p.mass=-0.2059381133516135
p.mass2=-0.04241050653082201
p.mass**2=0.042410506530822005
```

Well, the reconstructed vector is obviously unphysical and when I looked at the code I saw that `mass2` is an alias for `tau2` and it just returns just `E**2 - mag**2`, so this makes sense when the vector is unphysical and thus the magnitude is higher than the energy. I just wrongly assumed it would return the same as `mass**2` and at first I thought this breaks the principle of least surprise.

I looked at ROOT [`GenVector`](https://root.cern.ch/doc/master/GenVector_2LorentzVector_8h_source.html), which you try to be convention-compatible with, and it seems it does the same for `M2()` , so it seems this is not a bug but a feature and for ROOT users it keeps in line with expectations.. But `GenVector` seems to throw an exception when e.g. calculating `Beta()` with a negative `M2()`, while `vector` just quietly returns `NaN`'s.

**Suggestion:**
I think it would be nice to throw some exception when trying to boost with a time-like vector or at least print some warnings like numpy does when dividing by zero. And maybe have an optional keyword argument to disable exceptions. Then, the user would need to be more explicit to ignore these cases. Though I'm not sure if that's feasible from a design-perspective, maybe just returning NaN's is enough :shrug:

**Comment by jpivarski** [comment]
Considering that these are being applied to arrays of vectors, rather than an individual vector, raising exceptions is more problematic than setting NaN. (What if a huge dataset has just one such value, due to accumulated round-off?) That's why NumPy sets NaN—but with warnings (by default).

Vector suppresses these warnings because some of them are intermediate to the calculation:

https://github.com/scikit-hep/vector/blob/0b9bf7ef622b634b749a0fa664da2c2a565affa1/src/vector/_compute/lorentz/beta.py#L122-L133

But perhaps there should be a `vector.errstate` that does the same thing as NumPy's, but only when final results contain NaNs? Just deciding whether there are any NaNs in the output would require an additional pass over the data, with [np.isnan](https://numpy.org/doc/stable/reference/generated/numpy.isnan.html).
